lib/mayaUsd/ufe/UsdStageMap.cpp

```diff
@@ -230,21 +230,21 @@ UsdStageMap::StageSet UsdStageMap::allStages()
 {
     rebuildIfDirty();
 
-    // Note: stage() calls proxyShape() which can modify fPathToObject,
-    //       so we cannot call stage within a loop on fPathToObject.
-    //
-    //       So, we first cache all UFE paths and then loop over them
-    //       calling stage().
-
-    std::vector<Ufe::Path> paths;
-    paths.reserve(fPathToObject.size());
-    for (const auto& pair : fPathToObject) {
-        paths.emplace_back(pair.first);
-    }
-
     StageSet stages;
-    for (const auto& path : paths) {
-        stages.insert(stage(path));
+    for (auto it = fPathToObject.begin(); it != fPathToObject.end();
+         /* no ++it here, we manually move it in the loop */) {
+        const auto&      pair = *it;
+        const Ufe::Path& path = pair.first;
+        // Calling UsdStageMap::stage may erase the entry in fPathToObject at path.
+        // Advance the iterator so that the iterator stays valid if erasure occurs.
+        it++;
```

Comment:
Actually, with my new fixes, the entire collection might get rebuilt. I think this will break when my change go in, so I'd revert it to cache the paths like it was before.

Reply:
I'm not sure I agree with that (and neither did Krystian). Its very possible for a scene to contain thousands of proxy shapes, so rebuilding the cache just because one entry is invalid is inefficient. @pierrebai-adsk Your PR hasn't been approved yet so I don't think William should revert his changes just yet.